xray/run.py

```diff
@@ -0,0 +1,145 @@
+            return Wireguard.Native
+        elif s is not None and s.lower() == "boringtun":
+            return Wireguard.BoringTun
+        else:
+            raise Exception(f"{s} is not a valid wireguard type")
+
+
+def setup_wireguard(wg, build_neptun):
+    if wg == Wireguard.Native:
+        run_command(f"sudo ip link add dev {WG_IFC_NAME} type wireguard")
+    elif wg == Wireguard.WgGo:
+        wggo = (
+            run_command("which wireguard", capture_output=True)[0]
+            .strip()
+            .decode("utf-8")
+        )
+        run_command(f"sudo {wggo} {WG_IFC_NAME}")
+    elif wg == Wireguard.BoringTun:
+        run_command(f"sudo ../target/release/boringtun-cli {WG_IFC_NAME}")
+    else:
+        if build_neptun:
+            run_command(f"cargo run --release -p neptun-cli -- {WG_IFC_NAME}")
+        else:
+            run_command(f"sudo ../target/release/neptun-cli {WG_IFC_NAME}")
+    run_command(f"sudo ip link set dev {WG_IFC_NAME} up")
+    run_command(
+        f"sudo ip link set dev {WG_IFC_NAME} multicast off"
+    )  # Not strictly necessary but keeps the pcaps a bit cleaner
+
+
+def start_tcpdump(pcap_name):
+    return subprocess.Popen(
+        ["sudo", "tcpdump", "-ni", "any", "-w", pcap_name],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        text=True,
+    )
+
+
+def run_xray(wg, test_type, count):
+    run_command(
+        f"cargo run --release -- {wg} {test_type} {count} {get_csv_name(wg, test_type, count)}"
+    )
+
+
+def stop_tcpdump(tcpdump):
+    run_command(f"kill {tcpdump.pid}")
+
+
+def destroy_wireguard(wg):
+    if wg == Wireguard.NepTUN:
+        run_command("killall neptun-cli")
+    elif wg == Wireguard.BoringTun:
+        run_command("killall boringtun-cli")
+    else:
+        run_command(f"sudo ip link delete {WG_IFC_NAME}")
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--wg")
+    parser.add_argument("--test_type")
+    parser.add_argument("--count")
+    parser.add_argument("--nobuild", action="store_true")
+    args = parser.parse_args()
+
+    wg = Wireguard.from_str(args.wg)
+    test_type = args.test_type.lower() if args.test_type is not None else "crypto"
+    assert test_type in [
+        "crypto",
+        "pt",
+    ], f"Invalid test type '{test_type}'. Valid options are 'crypto' and 'pt'"
+    count = int(args.count) if args.count is not None else 10
+    assert count > 0, f"Count must be at least one, but got {count}"
+    build_neptun = args.nobuild is None
+
+    Path("results/").mkdir(parents=True, exist_ok=True)
+    try:
+        os.remove(get_csv_name(wg.name, test_type, count))
+        os.remove(get_pcap_name(wg.name, test_type, count))
+    except:  # noqa: E722
+        pass
+
+    setup_wireguard(wg, build_neptun)
+    tcpdump = start_tcpdump(get_pcap_name(wg.name, test_type, count))
+
+    succeeded = True
+    try:
+        run_xray(wg.name, test_type, count)
+    except:  # noqa: E722
```

Comment:
when does cargo run thrown an exception?

In case the test goes wrong, let's say the receiver doesn't get any packets, cargo process will still succeed which later will make `analyze()` fail.

I can't think right away of an alternative to check if we should proceed with analyze() but probably there is

Reply:
It was a simple as adding `check=True` to the `subprocess.run` command in my `run_command` function